dali/pipeline/operators/displacement/warp.h

```diff
@@ -0,0 +1,365 @@
+        type2id<typename FirstTypePair::first_type>::value == input_type_)
+      functor(FirstTypePair());
+    else
+      ToStaticTypeEx(std::tuple<TypePairs...>(), std::forward<F>(functor));
+  }
+
+  template <typename F>
+  void ToStaticType(F &&functor) {
+    using supported_types = typename MyType::SupportedTypes;
+    ToStaticTypeEx(supported_types(), std::forward<F>(functor));
+  }
+
+  // TODO(michalz): Change value switch over SpatialDim to (2, 3) when kernel is implemented
+  #define WARP_STATIC_TYPES(...) {                                          \
+    VALUE_SWITCH(This().SpatialDim(), spatial_ndim, (2), (                  \
+      VALUE_SWITCH(This().BorderClamp() ? 1 : 0, UseBorderClamp, (0, 1), (  \
+          ToStaticType(                                                     \
+            [&](auto &&args) {                                              \
+            using OutputType = decltype(args.first);                        \
+            using InputType = decltype(args.second);                        \
+            using BorderType = std::conditional_t<                          \
+              UseBorderClamp, kernels::BorderClamp, OutputType>;            \
+            __VA_ARGS__                                                     \
+          });),                                                             \
+          (assert(!"impossible")))),                                        \
+        (DALI_FAIL("Only 2D and 3D warping is supported")));                \
+  }
+
+  /// @}
+ public:
+  using Operator<Backend>::Operator;
+
+  int SpatialDim() const {
+    return input_shape_.sample_dim()-1;
+  }
+
+  bool BorderClamp() const {
+    return !Spec().HasArgument("border");
+  }
+
+  bool CanInferOutputs() const override {
+    return true;
+  }
+
+  bool SetupImpl(std::vector<OutputDesc> &outputs, const Workspace &ws) override {
+    outputs.resize(1);
+
+    DALIDataType out_type;
+    SetupWarp(outputs[0].shape, out_type, ws);
+    outputs[0].type = TypeTable::GetTypeInfo(out_type);
+    return true;
+  }
+
+  using DefaultSupportedTypes = UnzipPairs<
+    uint8_t, uint8_t,
+    uint8_t, float,
+    float,   uint8_t,
+
+    int16_t, int16_t,
+    int16_t, float,
+    float,   int16_t,
+
+    int32_t, int32_t,
+    int32_t, float,
+    float,   int32_t,
+
+    float,   float
+  >;
+
+  /// @brief May be shadowed by Derived, if necessary
+  using SupportedTypes = DefaultSupportedTypes;
+
+  void SetupWarp(kernels::TensorListShape<> &out_shape,
+                 DALIDataType &out_type,
+                 const Workspace &ws) {
+    auto &input = ws.template InputRef<Backend>(0);
+    input_shape_ = input.shape();
+    DALIDataType new_input_type = input.type().id();
+    DALIDataType new_output_type;
+    if (!Spec().TryGetArgument(new_output_type, "output_type"))
+      new_output_type = new_input_type;
+
+    output_type_ = new_output_type;
+    input_type_ = new_input_type;
+
+    WARP_STATIC_TYPES(
```

Comment:
I find this macro slightly over-complicated for the reader. I would rather like to see the whole code here

Reply:
Agreed, I thought I would need more than one use, but somehow managed with just one.